### PR TITLE
MGDOBR-1297: Refactoring error catalog in v1 and v2

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-1297](https://issues.redhat.com/browse/MGDOBR-1297)

This PR adds support for versions to use their own "Error Catalogue". (Almost) nothing is now shared.

Only `ExternalUserException` and `InternalPlatformException` are shared. Everything else is version dependent.

This led to the need to split some classes into a base (abstract) implementation with `v1` and `v2` sub-classes to correctly throw the correct `Exception`. There _could_ be an argument to move some more exceptions back to being "shared" to avoid this new split of a few classes. I tried to keep sharing to an absolute minimum to ensure maximum flexibility. 

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-XXX] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: infra-parent/infra-core/src/main/java/com/redhat/service/smartevents/infra/core/exceptions/ErrorHrefVersionBuilderImpl.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Inject
+    public ErrorHrefVersionBuilderImpl(@Any Instance<ErrorHrefVersionProvider> builders) {
+        this.builders = builders;
+    }
+
+    @Override
+    public String buildHref(Throwable e, String id) {
+        if (e instanceof HasErrorHref) {
```
Comment (manstis) [inline infra-parent/infra-core/src/main/java/com/redhat/service/smartevents/infra/core/exceptions/ErrorHrefVersionBuilderImpl.java:32]: Explicit RHOSE `Exceptions` have an explicit property for the `href`.

We now only use the `ErrorHrefVersionProvider` as a fallback mechanism for the generic exceptions `ExternalUserException` and `InternalPlatformException`. These need to be common to all versions as our single-instance/cross-version `ExceptionMapper` implementations need to use a common class.

**Comment by r00ta** [comment]
will take a look at this tomorrow :+1:

File: manager-parent/manager-core/src/test/java/com/redhat/service/smartevents/manager/core/dns/openshift/DnsServiceOpenshiftImplTest.java [hunk 2]
```diff
@@ -27,6 +29,8 @@
 import static org.mockito.Mockito.when;
 
 @QuarkusTest
+// This annotation also bootstraps Quarkus into using DevServices for ALL @QuarkusTest's
+@QuarkusTestResource(PostgresResource.class)
```
Comment (r00ta) [inline manager-parent/manager-core/src/test/java/com/redhat/service/smartevents/manager/core/dns/openshift/DnsServiceOpenshiftImplTest.java:33]: Is this on purpose or you were looking for `restrictToAnnotatedClass = true`?

**Review by r00ta — APPROVED** [review]

File: manager-parent/manager-core/src/test/java/com/redhat/service/smartevents/manager/core/dns/openshift/DnsServiceOpenshiftImplTest.java (hunk 2 above)
Comment (manstis) [reply manager-parent/manager-core/src/test/java/com/redhat/service/smartevents/manager/core/dns/openshift/DnsServiceOpenshiftImplTest.java:33]: Without the annotation Quarkus was complaining about there being an undefined DataSource.

I had moved a class from `manager-core` to `manager-v1` that had the annotation so I assumed its existence had allowed tests in `manager-core` to work previously. Adding the annotation to a(ny) test in `manager-core` reinstated successful execution of the tests.

Comment (manstis) [reply manager-parent/manager-core/src/test/java/com/redhat/service/smartevents/manager/core/dns/openshift/DnsServiceOpenshiftImplTest.java]: I dug a bit more...

It's needed for our DAO tests and Quartz.

I've moved the annotation from the Dns test to a DAO test where its inclusion makes more sense.

**Review by robpblake — APPROVED** [review]

**Review by winklerm — APPROVED** [review]

**Pull request merged**